### PR TITLE
Local subscriptions

### PR DESCRIPTION
**Pull request by gonzalojaubert**
## Description
This PR add **Local** subscriptions to Booster.

## Changes

* Local infrastructure needed for Subscriptions
* Local logic needed for Subscriptions

## Checks
- [x] Project Builds
- [X] Project passes tests and checks
- [X] Updated documentation accordingly

### CODE REVIEW

File: packages/framework-provider-local/src/library/graphql-adapter.ts [hunk 1]
```diff
@@ -1,45 +1,116 @@
...
+    const connectionContext = webSocketRequest.connectionContext
+    eventType = connectionContext?.eventType
     return {
-      requestID,
-      eventType,
-      connectionID,
-      token: request.headers.authorization,
-      value: request.body,
+      requestID: requestId,
+      eventType: eventType,
```
Comment (NickSeagull) [inline packages/framework-provider-local/src/library/graphql-adapter.ts:41]: You can actually rename `requestId` to `requestID` and make this like:
```suggestion
      requestID,
      eventType,
```

File: packages/framework-provider-local/src/library/web-socket-server-adapter.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      await this.graphQLService.handleGraphQLRequest(request)
+      this.onMessage(webSocket, connectionRequest, request.connectionContext.connectionId)
+      this.onDisconnect(webSocket, request.connectionContext.connectionId)
+    })
+  }
+
+  private onMessage(webSocket: WebSocket, connectionRequest: IncomingMessage, connectionId: UUID) {
+    webSocket.on('message', async (message) => {
+      const request: ExpressWebSocketMessage = {
+        data: JSON.parse(message.toString()),
```
Comment (NickSeagull) [inline packages/framework-provider-local/src/library/web-socket-server-adapter.ts:68]: This JSON.parse can actually fail in the case the string is not a proper JSON, let's better have a try/catch here

**Review by NickSeagull — CHANGES_REQUESTED** [review]

File: packages/framework-provider-local/src/library/web-socket-server-adapter.ts (hunk 2 above)
Comment (gonzalojaubert) [reply packages/framework-provider-local/src/library/web-socket-server-adapter.ts]: I update the code to ignore the error, but let me know what you think about it

File: packages/framework-provider-local/src/library/graphql-adapter.ts [hunk 3]
```diff
@@ -1,45 +1,116 @@
...
+        rawContext: httpRequest,
+      },
+    } as GraphQLRequestEnvelopeError
   }
 }
+
+function isExpressWebSocketMessage(
+  request: express.Request | ExpressWebSocketMessage
+): request is ExpressWebSocketMessage {
+  return Boolean((request as ExpressWebSocketMessage).connectionContext)
```
Comment (alvaroloes) [inline packages/framework-provider-local/src/library/graphql-adapter.ts:115]: You can use the `in` operator to simplify this:
```suggestion
  return "connectionContext" in request
```

File: packages/framework-provider-local/src/library/read-model-adapter.ts [hunk 4]
```diff
@@ -59,6 +60,12 @@ export async function storeReadModel(
     throw e
   }
   logger.debug('Read model stored')
+  try {
+    await graphQLService.handleNotificationSubscription([{ typeName: readModelName, value: readModel }])
+    logger.debug('Read model change notified')
+  } catch (e) {
+    logger.error('Error notifying subscription', readModel)
+  }
```
Comment (alvaroloes) [inline packages/framework-provider-local/src/library/read-model-adapter.ts:68]: Not for this PR, but it makes a lot of sense to add a SubscriptionNotificationError to the Global error handler, as we have with the reductions, projections, etc

File: packages/framework-provider-local/src/library/web-socket-server-adapter.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  private readonly httpServer: HttpServer
+  private readonly _webSocketServer: Server
+  private readonly clients: Record<string, WebSocket> = {}
+  private readonly WS_PORT = process.env[LOCAL_ENVIRONMENT_WEBSOCKET_SERVER_PORT] ?? '65529'
+
+  constructor(readonly graphQLService: GraphQLService, config: BoosterConfig) {
+    const expressServer = express()
+    this._webSocketServer = new WebSocket.Server({
+      noServer: true,
+      path: '/websockets',
```
Comment (alvaroloes) [inline packages/framework-provider-local/src/library/web-socket-server-adapter.ts:33]: Nitpicking: This should be singular, as it is the only websocket we create
```suggestion
      path: '/websocket',
```

File: packages/framework-provider-local/src/library/web-socket-server-adapter.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    logger.info(`Starting websocket server on ws://localhost:${this.WS_PORT}/websockets`)
+    this.attach()
+    this.onConnection(config)
+  }
+
+  public attach(): void {
+    this.httpServer.on('upgrade', (request: IncomingMessage, socket: Duplex, head: Buffer) => {
+      this._webSocketServer.handleUpgrade(request, socket, head, (webSocket: WebSocket) => {
+        this._webSocketServer.emit('connection', webSocket, request)
+      })
```
Comment (alvaroloes) [inline packages/framework-provider-local/src/library/web-socket-server-adapter.ts:46]: What a crazy thing XD

File: packages/framework-provider-local/src/library/web-socket-server-adapter.ts [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+      }
+      await this.graphQLService.handleGraphQLRequest(request)
+    })
+  }
+
+  public sendToConnection(connectionId: string, data: unknown): void {
+    const stringData = JSON.stringify(data)
+    this.clients[connectionId].send(stringData)
+  }
+}
```
Comment (alvaroloes) [inline packages/framework-provider-local/src/library/web-socket-server-adapter.ts:110]: Loved the code of this class 💯

File: packages/framework-provider-local/src/services/connection-registry.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+  constructor() {
+    this.connections.loadDatabase()
+  }
+
+  getCursor(query: object, createdAt = 1, projections?: unknown) {
+    return this.connections.find(query, projections).sort({ createdAt: createdAt })
+  }
+
+  public async query(query: object, createdAt = 1, limit?: number, projections?: unknown): Promise<unknown> {
```
Comment (alvaroloes) [inline packages/framework-provider-local/src/services/connection-registry.ts:20]: Why createdAt has a default value of 1?

File: packages/framework-provider-local/src/services/subscription-registry.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    const deletePromise = new Promise((resolve, reject) =>
+      this.subscriptions.remove({}, { multi: true }, (err, numRemoved: number) => {
+        if (err) reject(err)
+        else resolve(numRemoved)
+      })
+    )
+
+    return (await deletePromise) as number
+  }
+}
```
Comment (alvaroloes) [inline packages/framework-provider-local/src/services/subscription-registry.ts:61]: This class is (if I didn't miss something) exactly the same as ConnectionRegistry. We can have only one class (for example called SimpleRegistry) that has the database path as the only parameter. This way, we avoid duplicating all the code.

File: packages/framework-provider-local/src/index.ts [hunk 10]
```diff
@@ -20,17 +20,35 @@ import {
...
+  subscribeToReadModel,
+} from './library/subscription-adapter'
 
 export * from './paths'
 export * from './services'
 
 const eventRegistry = new EventRegistry()
 const readModelRegistry = new ReadModelRegistry()
+const connectionRegistry = new ConnectionRegistry()
+const subscriptionRegistry = new SubscriptionRegistry()
```
Comment (alvaroloes) [inline packages/framework-provider-local/src/index.ts:45]: If we apply the above suggestion, then we can do the following here:
```suggestion
const connectionRegistry = new SimpleRegistry(connectionsDatabase)
const subscriptionRegistry = new SimpleRegistry(subscriptionDatabase)
```

**Review by alvaroloes — COMMENTED** [review]
Amazing work! 🥇

File: packages/framework-provider-local/src/services/subscription-registry.ts (hunk 9 above)
Comment (gonzalojaubert) [reply packages/framework-provider-local/src/services/subscription-registry.ts:61]: Good point

File: packages/framework-provider-local/src/services/connection-registry.ts (hunk 8 above)
Comment (gonzalojaubert) [reply packages/framework-provider-local/src/services/connection-registry.ts]: It was a copy/paste issue

File: packages/framework-provider-local/src/library/read-model-adapter.ts (hunk 4 above)
Comment (gonzalojaubert) [reply packages/framework-provider-local/src/library/read-model-adapter.ts:68]: Good ideaw

File: packages/framework-provider-local/src/library/web-socket-server-adapter.ts [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+  connectionContext: ConnectionContext
+  incomingMessage?: IncomingMessage
+  data?: unknown
+}
+
+const LOCAL_ENVIRONMENT_WEBSOCKET_SERVER_PORT = 'LOCAL_ENVIRONMENT_WEBSOCKET_SERVER_PORT'
+
+export class WebSocketServerAdapter {
+  private readonly httpServer: HttpServer
+  private readonly _webSocketServer: Server
```
[the remark this answers is not in the document]
Comment (gonzalojaubert) [reply packages/framework-provider-local/src/library/web-socket-server-adapter.ts:25]: ¯\_(ツ)_/¯ I will fixed it :)

File: packages/framework-provider-local/src/library/subscription-adapter.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
+import { BoosterConfig, SubscriptionEnvelope } from '@boostercloud/framework-types'
+import { SubscriptionRegistry } from '../services/subscription-registry'
+
+export async function subscribeToReadModel(
+  db: SubscriptionRegistry,
+  config: BoosterConfig,
+  subscriptionEnvelope: SubscriptionEnvelope
+): Promise<void> {
+  await db.store(subscriptionEnvelope)
```
[the remark this answers is not in the document]
Comment (gonzalojaubert) [reply packages/framework-provider-local/src/library/subscription-adapter.ts:9]: No, I missed that part. Let me try to implement a solution for the LocalProvider

**Review by NickSeagull — APPROVED** [review]

File: packages/framework-provider-local/src/library/web-socket-server-adapter.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (alvaroloes) [reply packages/framework-provider-local/src/library/web-socket-server-adapter.ts]: jajajaja

**Review by alvaroloes — APPROVED** [review]

**Pull request merged**